pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,338 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"io"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+type PodSandboxID string
+
+// PodSandboxManager provides basic operations to create/delete and examine the
+// PodSandboxes.
+type PodSandboxManager interface {
+	// Create creates a sandbox based on the given config, and returns the
+	//  of the new sandbox.
+	Create(config *PodSandboxConfig) (PodSandboxID, error)
+	// Delete deletes the sandbox by its ID. If there are any running
+	// containers in the sandbox, they will be terminated as a side-effect.
+	Delete(id PodSandboxID) error
+	// List lists existing sandboxes, filtered by the given PodSandboxFilter.
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	// Status gets the status of the sandbox by ID.
+	Status(id PodSandboxID) (PodSandboxStatus, error)
+	// PortForward copies the data from the port in the PodSandBox to the
+	// stream.
+	// TODO: Should PortForward be in a separate interface.
+	PortForward(id PodSandboxID, port uint16, stream io.ReadWriteCloser) error
+}
+
+// PodSandboxConfig holds all the required and optional fields for creating a
+// sandbox.
+type PodSandboxConfig struct {
+	// Name is the name of the sandbox.
+	Name string
+	// Hostname is the hostname of the sandbox.
+	Hostname string
+	// DNSOptions sets the DNS options for the sandbox.
+	DNSOptions DNSOptions
+	// PortMappings lists the port mappings for the sandbox.
+	PortMappings []PortMapping
+	// Mounts lists the mounts to be added to the sandbox's filesystem.
+	Mounts []Mount
+	// Resources specifies the resource requirements for the sandbox.
+	// Note: On a Linux host, kubelet will create a pod-level cgroup and pass
+	// it as the cgroup parent for the PodSandbox. For some runtimes, this is
+	// sufficent to fulfill pod-level resource requirements, and this field
+	// will not be used. For others, e.g., hypervisor-based runtimes, explicit
+	// resource requirements for the sandbox are needed at creation time.
+	Resources PodSandboxResources
+	// Labels are key value pairs that may be used to scope and select individual resources.
+	Labels Labels
```

Comment:
Labels have to adhere to a standard schema. Applying such schema to
specific annotations keys or expecting runtimes to support grouping on
special annotation keys sounds a bit weird.

On Thu, May 26, 2016 at 3:15 PM, Euan Kemp <notifications@github.com> wrote:

> In pkg/kubelet/container/interface.go
> <https://github.com/kubernetes/kubernetes/pull/25899#discussion_r64830621>
> :
>
> > +	Hostname string
> > +	// DNSOptions sets the DNS options for the sandbox.
> > +	DNSOptions DNSOptions
> > +	// PortMappings lists the port mappings for the sandbox.
> > +	PortMappings []PortMapping
> > +	// Mounts lists the mounts to be added to the sandbox's filesystem.
> > +	Mounts []Mount
> > +	// Resources specifies the resource requirements for the sandbox.
> > +	// Note: On a Linux host, kubelet will create a pod-level cgroup and pass
> > +	// it as the cgroup parent for the PodSandbox. For some runtimes, this is
> > +	// sufficent to fulfill pod-level resource requirements, and this field
> > +	// will not be used. For others, e.g., hypervisor-based runtimes, explicit
> > +	// resource requirements for the sandbox are needed at creation time.
> > +	Resources PodSandboxResources
> > +	// Labels are key value pairs that may be used to scope and select individual resources.
> > +	Labels Labels
>
> That doesn't answer why labels can't or shouldn't just be a specialization
> of annotations (e.g. io.kuberentes.label/foo=bar) which is interpreted
> further
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25899/files/7674a3563e076840baf54e0ef6d24dd7b0c916b0#r64830621>
>